src/Data/StringVariants/Util.hs

```diff
@@ -23,3 +28,39 @@ usePositiveNat n a f = case someNatVal n of
     EQI -> f p
     LTI -> f p
     GTI -> a
+
+-- Logic ported from base-4.17 Data.Char.isSpace.
+-- Hopefully, GHC would add this type-level function
+type IsCharWhitespace c
+  = If
+    (CharToNat c <=? 0x376)
+    (CharToNat c == 32 || CharToNat c - 0x9 <=? 4 || CharToNat c == 0xa0)
+    (IsUnicodeWhitespaceCodePoint (CharToNat c))
+
+-- It checks the whitespace characters (Unicode property White_Space=yes) defined in Unicode 15.0.0 and above the ASCII range.
+type IsUnicodeWhitespaceCodePoint n
+  = n == 0x1680
+  || (0x2000 <=? n && n <=? 0x200A)
+  || n == 0x2028 || n == 0x2029 || n == 0x202F || n == 0x205F || n == 0x3000
+
+type family SymbolLength (length :: Nat) (s :: Maybe (Char, Symbol)) :: Nat where
+  SymbolLength length 'Nothing = length
+  SymbolLength length ('Just '(_, s)) = SymbolLength (length + 1) (UnconsSymbol s)
+
+type SymbolNonEmpty s = If (s == "") (TypeError ('Text "Symbol is empty")) (() :: Constraint)
+
+type family SymbolNoLeadingSpace (s :: Maybe (Char, Symbol)) :: Constraint where
+  SymbolNoLeadingSpace 'Nothing = ()
+  SymbolNoLeadingSpace ('Just '(c, _)) = If (IsCharWhitespace c) (TypeError ('Text "Symbol has leading whitespace")) (() :: Constraint)
+
+type family SymbolNoTrailingSpace (s :: Maybe (Char, Symbol)) :: Constraint where
+  SymbolNoTrailingSpace 'Nothing = ()
+  SymbolNoTrailingSpace ('Just '(c, "")) = If (IsCharWhitespace c) (TypeError ('Text "Symbol has trailing whitespace")) (() :: Constraint)
+  SymbolNoTrailingSpace ('Just '(_, s)) = SymbolNoTrailingSpace (UnconsSymbol s)
+
+-- The type family IsCharWhitespace is large but it is only evaluated for the first and last symbols
+type SymbolWithNoSpaceAround s = (SymbolNoLeadingSpace (UnconsSymbol s), SymbolNoTrailingSpace (UnconsSymbol s)) 
+
+type SymbolNoLongerThan s n = If (SymbolLength 0 (UnconsSymbol s) <=? n)
+  (() :: Constraint)
+  (TypeError ('Text "Invalid NonEmptyText. Needs to be <= " ':<>: 'ShowType n ':<>: 'Text " characters. Has " ':<>: 'ShowType (SymbolLength 0 (UnconsSymbol s)) ':<>: 'Text " characters."))
```

Comment:
I feel like this should be rewritten to preserve laziness, especially since we're recomputing the `SymbolLength 0 (UnconsSymbol s)`. 

```suggestion
type SymbolNoLongerThan s n =
  SymbolNoLongerThan' n (SymbolLength 0 (UnconsSymbol s))
 
type family SymbolNoLongerThanImpl s n len where
  SymbolNoLongerThanImpl n len =
    If (len <=? n) 
      (() :: Constraint) 
      (TypeError ('Text "Invalid NonEmptyText. Needs to be <= " ':<>: 'ShowType n ':<>: 'Text " characters. Has " ':<>: 'ShowType len) ':<>: 'Text " characters."))
```

Another inline defun would rid the `If`, but since the evaluations are pretty simple, it's probably fine.